components/style/gecko_bindings/sugar/ownership.rs

```diff
@@ -108,140 +108,17 @@ pub unsafe trait HasArcFFI : HasFFI {
             transmute::<&&Self::FFIType, &Arc<Self>>(ptr)
         }
     }
-}
-
-#[repr(C)]
-/// Gecko-FFI-safe borrowed type
-/// This can be null.
-pub struct Borrowed<'a, T: 'a> {
-    ptr: *const T,
-    _marker: PhantomData<&'a T>,
-}
-
-#[repr(C)]
-/// Gecko-FFI-safe mutably borrowed type
-/// This can be null.
-pub struct BorrowedMut<'a, T: 'a> {
-    ptr: *mut T,
-    _marker: PhantomData<&'a mut T>,
-}
-
-// manual impls because derive doesn't realize that `T: Clone` isn't necessary
-impl<'a, T> Copy for Borrowed<'a, T> {}
-
-impl<'a, T> Clone for Borrowed<'a, T> {
-    #[inline]
-    fn clone(&self) -> Self { *self }
-}
-
-impl<'a, T> Borrowed<'a, T> {
-    #[inline]
-    pub fn is_null(self) -> bool {
-        self.ptr == ptr::null()
-    }
 
     #[inline]
-    /// Like Deref, but gives an Option
-    pub fn borrow_opt(self) -> Option<&'a T> {
-        if self.is_null() {
-            None
-        } else {
-            Some(unsafe { &*self.ptr })
-        }
-    }
-
-    #[inline]
-    /// Borrowed<GeckoType> -> Option<&Arc<ServoType>>
-    pub fn as_arc_opt<U>(&self) -> Option<&Arc<U>> where U: HasArcFFI<FFIType = T> {
+    fn arc_from_borrowed<'a>(ptr: &'a Option<&Self::FFIType>) -> Option<&'a Arc<Self>> {
```

Comment:
Nope. That's a dangling lifetime. `&Self::FFIType` corresponds to the arc, but that pointer is stack data which may be invalidated. We need to borrow the pointer itself.

This issue is super tricky and I've fallen into that trap in the past. Tried to be careful this time 😄